### PR TITLE
Gui menu mnemonic and accelerator key proposal

### PR DESCRIPTION
**Pull request by PPAC37**
I like being able to use keyboard shortcuts...
We must agree on which for what ...

### CODE REVIEW

**Comment by PPAC37** [comment]
I also moved the MenuItem Check For update to the Help Menu and added a Make a PayPal Donation menu item to the Help menu. (this too to discuss.)

File: src/main/java/com/marginallyclever/makelangelo/Makelangelo.java [hunk 1]
```diff
@@ -553,6 +571,8 @@ private JMenu createFileMenu() {
 				WindowEvent windowClosing = new WindowEvent(mainFrame, WindowEvent.WINDOW_CLOSING);
 				Toolkit.getDefaultToolkit().getSystemEventQueue().postEvent(windowClosing);
 			});
+			buttonExit.setAccelerator(KeyStroke.getKeyStroke("ctrl Q"));
+			buttonExit.setMnemonic('q');
```
Comment (gmembre-zenika) [inline src/main/java/com/marginallyclever/makelangelo/Makelangelo.java:575]: this one is to easy to type by mistake

**Comment by gmembre-zenika** [comment]
very nice !

**Comment by coliss86** [comment]
Very nice !

File: src/main/java/com/marginallyclever/makelangelo/Makelangelo.java (hunk 1 above)
Comment (coliss86) [inline src/main/java/com/marginallyclever/makelangelo/Makelangelo.java:575]: this one is too easy to hit by mistake

Comment (PPAC37) [reply src/main/java/com/marginallyclever/makelangelo/Makelangelo.java:575]: the mnemonic 'q' (only active if in the File menu (in used to do alt + f , q to quit some application ) ) 
or the Accelerator "ctrl Q" (all the time active)
or both ?

Maybe no mnemonic and just eventualy an accelerator "alt F4" (as this is windows default to quit, just so the user have a reminder ) ?

Comment (coliss86) [reply src/main/java/com/marginallyclever/makelangelo/Makelangelo.java]: My bad, I miss understood, leave it as you have first done !

File: src/main/java/com/marginallyclever/makelangelo/Makelangelo.java [hunk 2]
```diff
@@ -314,19 +316,22 @@ public void windowClosing(WindowEvent e) {
 
 	private JMenu createRobotMenu() {
 		JMenu menu = new JMenu(Translator.get("Robot"));
-
+		menu.setMnemonic('k');
 		menu.add(createRobotStyleMenu());
 		
 		JMenuItem bEstimate = new JMenuItem(Translator.get("RobotMenu.GetTimeEstimate"));
+		bEstimate.setAccelerator(KeyStroke.getKeyStroke("ctrl E"));
```
Comment (i-make-robots) [inline src/main/java/com/marginallyclever/makelangelo/Makelangelo.java:323]: KeyStroke.getKeyStroke(KeyEvent.VK_E, ActionEvent.CTRL_MASK)

Comment (i-make-robots) [reply src/main/java/com/marginallyclever/makelangelo/Makelangelo.java:323]: https://docs.oracle.com/javase/tutorial/uiswing/components/menu.html

File: src/main/java/com/marginallyclever/makelangelo/Makelangelo.java [hunk 3]
```diff
@@ -529,21 +543,25 @@ private JMenu createFileMenu() {
...
 		JMenuItem buttonAdjustPreferences = new JMenuItem(Translator.get("MenuPreferences"));
 		buttonAdjustPreferences.addActionListener((e)-> myPreferencesPanel.run(mainFrame));
+		buttonAdjustPreferences.setAccelerator(KeyStroke.getKeyStroke("alt P"));
 		menu.add(buttonAdjustPreferences);
 
 		JMenuItem buttonFirmwareUpdate = new JMenuItem(Translator.get("FirmwareUpdate"));
 		buttonFirmwareUpdate.addActionListener((e) -> runFirmwareUpdate());
 		menu.add(buttonFirmwareUpdate);
 		
+		/* // PPAC37 : moved in Help Menu.
```
Comment (i-make-robots) [inline src/main/java/com/marginallyclever/makelangelo/Makelangelo.java:560]: if this is moved then take away this dead code, please.

File: src/main/java/com/marginallyclever/makelangelo/Makelangelo.java [hunk 4]
```diff
@@ -658,7 +652,7 @@ private JMenu createViewMenu() {
 		menu.add(buttonZoomToFit);
 
 		JCheckBoxMenuItem checkboxShowPenUpMoves = new JCheckBoxMenuItem(Translator.get("GFXPreferences.showPenUp"),GFXPreferences.getShowPenUp());
-		checkboxShowPenUpMoves.setAccelerator(KeyStroke.getKeyStroke("shift M"));
```
Comment (i-make-robots) [inline src/main/java/com/marginallyclever/makelangelo/Makelangelo.java]: imo shift should never been an accelerator key.  I'm thinking of "your message here" where you can type in any language and that will eventually have someone type shift+letter to start a sentence.

File: src/main/java/com/marginallyclever/makelangelo/Makelangelo.java [hunk 5]
```diff
@@ -450,15 +456,20 @@ private JMenu createToolsMenu() {
 		menu.addSeparator();
 		
 		TurtleModifierAction a4 = new FlipTurtleAction(1,-1,Translator.get("FlipH"));
+		//a4.putValue(Action.ACCELERATOR_KEY, KeyStroke.getKeyStroke(java.awt.event.KeyEvent.VK_H, java.awt.event.InputEvent.SHIFT_DOWN_MASK));//"shift H" //shift should never been an accelerator key
```
Comment (coliss86) [inline src/main/java/com/marginallyclever/makelangelo/Makelangelo.java:459]: it is dead code, isn't it ? 
If yes, you should remove it :)

Comment (PPAC37) [reply src/main/java/com/marginallyclever/makelangelo/Makelangelo.java]: Ha yes it has become dead ... the time I find another less problematic keyboard shortcuts.
Thanks this comment got me out of procrastination. :)

**Comment by i-make-robots** [comment]
Let me know when you are happy with this and I will merge it.  Seems fine to me now.

**Comment by PPAC37** [comment]
For me this seem good ...

@coliss86  ?

|class|name|text|mnemo|accelerator|
|-|-|-|-|-|
|`javax.swing.JMenu                       `|`null                `|`File                          `|F|
|&nbsp;`javax.swing.JPopupMenu                 `|||||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`New                           `||ctrl pressed N|
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Open File...                  `||ctrl pressed O|
|&nbsp;&nbsp;`com.marginallyclever.makelangelo.RecentFiles`|`null                `|`Reopen last file              `||
|&nbsp;&nbsp;&nbsp;`javax.swing.JPopupMenu               `|||||
|&nbsp;&nbsp;&nbsp;&nbsp;`javax.swing.JMenuItem               `|`null                `|`/home/.../b4c8.dxf`||
|&nbsp;&nbsp;&nbsp;&nbsp;`javax.swing.JMenuItem               `|`null                `|`/home/.../screenshot.png`||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Save File...                  `||ctrl pressed S|
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Preferences                   `||alt pressed P|
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Update firmware               `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Quit                          `|Q|ctrl pressed Q|
|`javax.swing.JMenu                       `|`null                `|`Settings                      `|S|
|&nbsp;`javax.swing.JPopupMenu                 `|||||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Paper settings                `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Plotter settings              `||ctrl pressed P|
|`javax.swing.JMenu                       `|`null                `|`Generate art                  `|A|
|&nbsp;`javax.swing.JPopupMenu                 `|||||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Outline paper                 `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Dragon Fractal                `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Fibonacci Spiral Fractal      `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Fill page                     `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Gosper curve                  `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Graph Paper                   `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Hilbert Curve Fractal         `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Koch Curve Fractal            `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Lissajous                     `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`L System Tree Fractal         `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Maze                          `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Papercraft Box                `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Papercraft Polyhedron         `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Sierpinski triangle fractal   `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Spirograph                    `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Your message here             `||
|`javax.swing.JMenu                       `|`null                `|`Tools                         `|T|
|&nbsp;`javax.swing.JPopupMenu                 `|||||
|&nbsp;&nbsp;`javax.swing.JMenu$1                   `|`null                `|`Fit image on paper            `||
|&nbsp;&nbsp;`javax.swing.JMenu$1                   `|`null                `|`Fill entire paper             `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Scale                         `||
|&nbsp;&nbsp;`javax.swing.JMenu$1                   `|`null                `|`Flip horizontal               `||ctrl pressed H|
|&nbsp;&nbsp;`javax.swing.JMenu$1                   `|`null                `|`Flip vertical                 `||ctrl pressed F|
|&nbsp;&nbsp;`javax.swing.JMenu$1                   `|`null                `|`Simplify                      `||ctrl pressed Y|
|&nbsp;&nbsp;`javax.swing.JMenu$1                   `|`null                `|`Reorder                       `||ctrl pressed R|
|&nbsp;&nbsp;`javax.swing.JMenu$1                   `|`null                `|`Fill in closed loops          `||ctrl pressed I|
|`javax.swing.JMenu                       `|`null                `|`View                          `|V|
|&nbsp;`javax.swing.JPopupMenu                 `|||||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Zoom -                        `|-|ctrl pressed MINUS|
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Zoom +                        `|=|ctrl pressed EQUALS|
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Zoom to fit                   `|0|ctrl pressed 0|
|&nbsp;&nbsp;`javax.swing.JCheckBoxMenuItem         `|`null                `|`Show travel moves             `||ctrl pressed M|
|&nbsp;&nbsp;`javax.swing.JMenu                     `|`null                `|`Render style                  `||
|&nbsp;&nbsp;&nbsp;`javax.swing.JPopupMenu               `|||||
|&nbsp;&nbsp;&nbsp;&nbsp;`javax.swing.JRadioButtonMenuItem    `|`null                `|`Default                       `||
|&nbsp;&nbsp;&nbsp;&nbsp;`javax.swing.JRadioButtonMenuItem    `|`null                `|`Barber pole                   `||
|&nbsp;&nbsp;&nbsp;&nbsp;`javax.swing.JRadioButtonMenuItem    `|`null                `|`Separate loops                `||
|&nbsp;&nbsp;&nbsp;&nbsp;`javax.swing.JRadioButtonMenuItem    `|`null                `|`Marlin simulation             `||
|`javax.swing.JMenu                       `|`null                `|`Makelangelo                   `|K|
|&nbsp;`javax.swing.JPopupMenu                 `|||||
|&nbsp;&nbsp;`javax.swing.JMenu                     `|`null                `|`Robot style                   `||
|&nbsp;&nbsp;&nbsp;`javax.swing.JPopupMenu               `|||||
|&nbsp;&nbsp;&nbsp;&nbsp;`javax.swing.JRadioButtonMenuItem    `|`null                `|`Makelangelo 5                 `||
|&nbsp;&nbsp;&nbsp;&nbsp;`javax.swing.JRadioButtonMenuItem    `|`null                `|`Makelangelo 3.3               `||
|&nbsp;&nbsp;&nbsp;&nbsp;`javax.swing.JRadioButtonMenuItem    `|`null                `|`Makelangelo (Custom)          `||
|&nbsp;&nbsp;&nbsp;&nbsp;`javax.swing.JRadioButtonMenuItem    `|`null                `|`Cartesian                     `||
|&nbsp;&nbsp;&nbsp;&nbsp;`javax.swing.JRadioButtonMenuItem    `|`null                `|`Zarplotter                    `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Estimate time                 `||ctrl pressed E|
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Save Gcode to file/SD         `||ctrl pressed G|
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Open Controls                 `||ctrl pressed C|
|`javax.swing.JMenu                       `|`null                `|`Help                          `|H|
|&nbsp;`javax.swing.JPopupMenu                 `|||||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Show Log                      `||ctrl pressed L|
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Online help forums            `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`PayPal donation               `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`Check for updates             `||
|&nbsp;&nbsp;`javax.swing.JMenuItem                 `|`null                `|`About                         `||

**Comment by coliss86** [comment]
I love this feature !
All my comments has been taken in account, nothing more from me :)

File: src/main/java/com/marginallyclever/makelangelo/Makelangelo.java [hunk 6]
```diff
@@ -261,13 +261,15 @@ private void enableMenuBar(boolean b) {
 
 	private JMenu createSettingsMenu() {
 		JMenu menu = new JMenu(Translator.get("MenuSettings"));
+		menu.setMnemonic('S');
```
Comment (i-make-robots) [inline src/main/java/com/marginallyclever/makelangelo/Makelangelo.java:264]: I'm not a fan of single key mnemonics, same reason as using shift.  Change my mind!

Comment (PPAC37) [reply src/main/java/com/marginallyclever/makelangelo/Makelangelo.java:264]: You may be mixing concept of mnemonic key with the accelerators ?
normally, mnemonics key are active only if in the menu, (F10 or "alt" pressed to active the tool bar) ( eg : "alt f" then "q" to quit ). 
(and also dont know how to make a not single key mnemonics ?  )

**Comment by i-make-robots** [comment]
please resolve conflicts and i will merge.

**Comment by i-make-robots** [comment]
nice!

**Comment by PPAC37** [comment]
Except I messed up the indentation ... I'll do a PR to fix this tomorrow.

**Pull request merged**